rfcs/rfc-0000-interactions.md

```diff
@@ -0,0 +1,599 @@
+        date: String,
+        time: u32,
+    },
+    Attachment {
+        value: String,
+    },
+    VarArgs {
+        values: Vec<InteractionDataSlashCommandOptionInner>,
+    },
+    Group {
+        group: InteractionDataSlashCommandOption,
+    },
+}
+```
+
+## Responding To Interactions
+
+After receiving an interaction the bot has 3 seconds to respond to the interaction with a reponse, this
+response can be either a message if the interaction is attached to a channel or a modal.
+
+As not all validation can happen client side and api side the response can be an error response to indicate
+something failed validation or some other error for the user, this can include a message to the user to
+indicate why it failed, it can also optionally reenter the data they entered to make it easier to change the
+data the user typed in and send correct data.
+
+If the interaction is attached to a channel the interaction will also include a webhook which can be used
+to send followup messages, this webhook will last for 15 minuites then expire.
+
+## Config Panel
+
+The config panel structure is fetched everytime it is accessed, unlike slash commands where they must be synced,
+this is because it can include a lot of dynamic data like defaults - which should be the existing config values
+if any, select choices which use any members, roles, channels or similar from the context of where the panel is
+being used.
+
+There will be a single save button which initiats the interaction flow for the config panel, like other
+interaction types, this can lead into new interactions like modals, this can also respond with a failed validation
+response indicating that the values where wrong or some other error.
+
+There are two interaction flows present in config panels:
+
+- A user opens the panel, the API then must initiate a interaction flow to fetch the config panel, this is then
+displayed on the client to the user.
+
+- A user edits and saves the values in the config panel, this initiate a interaction flow to forward the config
+values to the bot to be saved.
+
+## Dashboard
+
+todo
+
+## Routes
+
+## Slash Command Management
+
+These routes are for managing slash commands on the bot
+
+```http
+GET /interactions/slash_commands - Get interactions
+PUT - /interactions/slash_commands - Sync interaction
+```
+
+## Config Panel
+
+
+These routes are the server and channel bot config panels
+
+```http
+GET /interactions/config_panel - Get server/channel config - spawns interactions
+PATCH /interactions/config_panel - Update server/channel config
+PATCH /interactions/config_panel/invoke - Set server/channel config options - spawns interactions
+```
+
+## Responding To Interactions
+
+These routes are for responding to interactions, this does not include the webhook routes for followups as those
+are already defined.
+
+```http
+POST /interactions/response - interaction response
+PATCH /interactions/response - edit interaction response
+```
+
+# Drawbacks
+
+todo
```

Comment:
Personally speaking, I believe one drawback could be that when multiple bots share the same command name (i.e. `/help`) that it could be a bit more tedious to find the command of the right bot, and much easier to missclick and select the wrong command from the wrong bot.

Reply:
One way to prevent dangerous command (due to missclick) is to confirm the user about their action, we can let the developer decide whether a command is dangerous or not.  Let's say `/ban @user1 @user2`, and we have two different bots implementing same command. We just confirm whether he is aware of the consequences. I don't know how feasible this will be!

Reply:
A general confirmation for possibly dangerous commands would indeed be good.
But maybe instead of having the dev tell the API whether this command is dangerous, it should be determined by something like permissions required to use the command (Assuming permissions for command usage will be a thing).
If the command required `Ban` permission, then you could safely assume it's a dangerous one and send the command confirmation to the user.

Also, my point wasn't really in regards to dangerous commands, but just general misclicks one could make when typing the commands.

Reply:
Sorry I misunderstood you, we can avoid misclicks by letting the user select the bot, but it could be a problem since if the user does not know which bot to use (this is really rare since the users who use commands are mostly mods who add the bots). Either way, I think letting the user select the bot before showing commands could be an alternative.